File: custom-rules/web-vulns/xpath-injection.test.py
# Test cases for python-xpath-injection rule

from flask import request
from lxml import etree
import re

# =============================================================================
# TRUE POSITIVES - Should be detected
# =============================================================================

# ruleid: python-xpath-injection
@app.route('/login', methods=['POST'])
def login_vulnerable():
    username = request.form['username']
    password = request.form['password']
    tree = etree.parse('users.xml')
    query = f"//user[username='{username}' and password='{password}']"
    result = tree.xpath(query)
    return "OK" if result else "Fail"

# ruleid: python-xpath-injection
@app.route('/search')
def search_vulnerable():
    name = request.args.get('name')
    tree = etree.parse('data.xml')
    result = tree.xpath("//item[name='" + name + "']")
    return str(result)

# ruleid: python-xpath-injection
@app.route('/user')
def user_vulnerable():
    user_id = request.args['id']
    doc = etree.parse('users.xml')
    expr = etree.XPath(f"//user[@id='{user_id}']")
    return str(expr(doc))

# =============================================================================
# TRUE NEGATIVES - Should NOT be detected
# =============================================================================

# ok: python-xpath-injection
def static_query():
    # Safe: Hardcoded query, no user input
    tree = etree.parse('data.xml')
    result = tree.xpath("//user[@admin='true']")
    return result

# ok: python-xpath-injection
def hardcoded_value():
    # Safe: Query built from hardcoded strings only
    tree = etree.parse('config.xml')
    field = "status"
    value = "active"
    result = tree.xpath(f"//item[@{field}='{value}']")
    return result

# ok: python-xpath-injection
def integer_path_variable():
    # Safe: Path variable that's cast to int immediately, then used as int
    tree = etree.parse('data.xml')
    item_id = 42
    result = tree.xpath(f"//item[@id='{item_id}']")
    return result

File: custom-rules/web-vulns/xpath-injection.test.php
<?php
// Test cases for php-xpath-injection rule

// =============================================================================
// TRUE POSITIVES - Should be detected
// =============================================================================

// ruleid: php-xpath-injection
function login_vulnerable() {
    $username = $_POST['username'];
    $password = $_POST['password'];
    $xml = simplexml_load_file('users.xml');
    $result = $xml->xpath("//user[username='$username' and password='$password']");
    return count($result) > 0;
}

// ruleid: php-xpath-injection
function search_vulnerable() {
    $name = $_GET['name'];
    $xml = simplexml_load_file('products.xml');
    $query = "//product[name='" . $name . "']";
    return $xml->xpath($query);
}

// ruleid: php-xpath-injection
function domxpath_vulnerable() {
    $id = $_REQUEST['id'];
    $doc = new DOMDocument();
    $doc->load('data.xml');
    $xpath = new DOMXPath($doc);
    return $xpath->query("//item[@id='$id']");
}

// ruleid: php-xpath-injection
function laravel_vulnerable(Request $request) {
    $category = $request->input('category');
    $xml = simplexml_load_string($data);
    return $xml->xpath("//product[@category='$category']");
}

// =============================================================================
// TRUE NEGATIVES - Should NOT be detected
// =============================================================================

// ok: php-xpath-injection
function static_query() {
    // Safe: Hardcoded query
    $xml = simplexml_load_file('users.xml');
    return $xml->xpath("//user[@admin='true']");
}

// ok: php-xpath-injection
function hardcoded_values() {
    // Safe: Query with hardcoded values only
    $xml = simplexml_load_file('config.xml');
    $field = "status";
    $value = "active";
    return $xml->xpath("//setting[@$field='$value']");
}

// ok: php-xpath-injection
function literal_query() {
    // Safe: All literals, no variables at all
    $xml = simplexml_load_file('data.xml');
    return $xml->xpath("//item[@type='product' and @active='1']");
}
?>
